### PR TITLE
Change `--max-imbalance-fee` to `--proportional-imbalance-fee`

### PR DESCRIPTION
**Pull request by palango**
Fixes: https://github.com/raiden-network/raiden/issues/4654

## Description

This makes the definition of the imbalance penalty fee relative to the
capacity to the channel. The goal is to make the setting more useful
until we implement a endpoint to configure it in the API.

## PR review check list

Quality check list that cannot be automatically verified.

- Safety
    - [ ] The changes respect the necessary conditions for safety (https://raiden-network-specification.readthedocs.io/en/latest/smart_contracts.html#protocol-values-constraints)
-  Code quality
    - [ ] Error conditions are handled
    - [ ] Exceptions are propagated to the correct parent greenlet
    - [ ] Exceptions are correctly classified as recoverable or unrecoverable
- Compatibility
    - [ ] State changes are forward compatible
    - [ ] Transport messages are backwards and forward compatible
- Commits
    - [ ] Have good messages
    - [ ] Squashed unecessary commits
- If it's a bug fix:
    - Regression test for the bug
        - [ ] Properly covers the bug
        - [ ] If an integration test is used, it could not be written as a unit test
- Is it a user facing feature/bug fix?
    - [ ] Changelog entry has been added

### CODE REVIEW

**Comment by pirapira** [comment]
Who pays the imbalance fee to whom?

**Comment by pirapira** [comment]
I cannot find imbalance fees in the spec https://github.com/raiden-network/spec/issues/251

**Comment by palango** [comment]
> I cannot find imbalance fees in the spec [raiden-network/spec#251](https://github.com/raiden-network/spec/issues/251)

Good point, we nbeed to add this. However, here is an ADR: https://github.com/raiden-network/raiden-services/blob/master/adr/003-mediation-fees.md

File: raiden/utils/typing.py [hunk 1]
```diff
@@ -103,6 +103,9 @@ def typecheck(value: Any, expected: Type):
 T_FeeAmount = int
 FeeAmount = NewType("FeeAmount", T_FeeAmount)
 
+T_RelativeFeeAmount = int
+RelativeFeeAmount = NewType("RelativeFeeAmount", T_RelativeFeeAmount)
```
Comment (pirapira) [inline raiden/utils/typing.py]: This is a good place to document the reasonable range of values, and what they mean.

Maybe `1` means one-per-million?  Or not?  It's not easy to figure out.

File: raiden/tests/unit/transfer/mediated_transfer/test_mediation_fee.py [hunk 2]
```diff
@@ -71,24 +77,28 @@ def test_linspace():
...
+    assert max(y for _, y in sample) == 1  # 20% of the 5 TA per channel side
 
-    sample = calculate_imbalance_fees(TA(1), max_imbalance_fee)
+    sample = calculate_imbalance_fees(TA(1), RFA(1_000_000))  # 100%
     assert sample is not None
     assert len(sample) == 2
+    assert all(0 <= x <= 1 for x, _ in sample)
     assert max(x for x, _ in sample) == 1
-    assert max(y for _, y in sample) == 10 ** 18
+    assert all(y == 0 for _, y in sample)
```
Comment (pirapira) [inline raiden/tests/unit/transfer/mediated_transfer/test_mediation_fee.py]: It's hard for me to expect all zero here after seeing 100%.  Maybe I'll figure out why after reading more.

File: raiden/transfer/mediated_transfer/mediation_fee.py [hunk 3]
```diff
@@ -93,18 +93,24 @@ def linspace(start: TokenAmount, stop: TokenAmount, num: int) -> List[TokenAmoun
...
 ) -> Optional[List[Tuple[TokenAmount, FeeAmount]]]:
     """ Calculates a quadratic rebalancing curve.
 
-    The penalty term takes the value `max_imbalance_fee` at the extrema.
+    The penalty term takes the following value at the extrema:
+        (channel_capacity / 2) * (proportional_imbalance_fee / int(1e6)
     """
-    if max_imbalance_fee == 0:
+    if proportional_imbalance_fee == 0:
         return None
```
Comment (pirapira) [inline raiden/transfer/mediated_transfer/mediation_fee.py]: I don't know when this function is supposed to return `None`.  Here it seems possible to return `[(somethihng, 0), (something, 0), ...]`.  Maybe the docstring can explain when it's `None`?

Comment (palango) [reply raiden/transfer/mediated_transfer/mediation_fee.py]: In that case it's true, we could just return two points as you write above. However in the case `channel_capacity == 0:` we don't support a return list with only one element, so we chose to return `None`  in this case.

File: raiden/tests/unit/transfer/mediated_transfer/test_mediation_fee.py (hunk 2 above)
Comment (palango) [reply raiden/tests/unit/transfer/mediated_transfer/test_mediation_fee.py]: So this is a rounding issue. I'm not sure what to do. I guess in general the case with 1 token wei is not common, but I wanted it to be stated explicitly in the test.

- `floor` and `int` and `round` give the same result
- `ceil` would return 1.

I guess it's best to return the division by two above, to make this better in this extreme case.

File: raiden/transfer/mediated_transfer/mediation_fee.py [hunk 4]
```diff
@@ -93,18 +93,24 @@ def linspace(start: TokenAmount, stop: TokenAmount, num: int) -> List[TokenAmoun
...
 
     if channel_capacity == 0:
         return None
 
+    # calculate the maximum imbalance fee for the channel. As the imbalance penalty
+    # function is currently a quadratic function centered around the channel of both deposits,
+    # we have to divide by 2 here.
+    max_imbalance_fee = int(channel_capacity * proportional_imbalance_fee / int(1e6) / 2)
```
[the remark this answers is not in the document]
Comment (pirapira) [reply raiden/transfer/mediated_transfer/mediation_fee.py]: @pirapira ask on Stackoverflow whether `A * B` requires `A` and `B` to be of the same type.

File: raiden/ui/cli.py [hunk 5]
```diff
@@ -412,14 +411,17 @@ def options(func):
                 "--proportional-fee",
                 help="Mediation fee as ratio of mediated amount in parts-per-million (10^-6).",
                 default=DEFAULT_MEDIATION_PROPORTIONAL_FEE,
-                type=int,
+                type=click.IntRange(min=0),
```
Comment (pirapira) [inline raiden/ui/cli.py]: Also set `max=10**6`?

Comment (palango) [reply raiden/ui/cli.py]: Should we limit this? Why not allow 2*10**18?

File: raiden/transfer/mediated_transfer/mediation_fee.py [hunk 6]
```diff
@@ -93,18 +93,24 @@ def linspace(start: TokenAmount, stop: TokenAmount, num: int) -> List[TokenAmoun
...
-    if max_imbalance_fee == 0:
+    if proportional_imbalance_fee == 0:
         return None
 
     if channel_capacity == 0:
         return None
 
+    # calculate the maximum imbalance fee for the channel. As the imbalance penalty
+    # function is currently a quadratic function centered around the channel of both deposits,
+    # we have to divide by 2 here.
```
[the remark this answers is not in the document]
Comment (palango) [reply raiden/transfer/mediated_transfer/mediation_fee.py]: Good point, this was contradicting to the description of the cli argument. I removed the division by  two.

File: raiden/tests/unit/transfer/mediated_transfer/test_mediation_fee.py (hunk 2 above)
Comment (pirapira) [reply raiden/tests/unit/transfer/mediated_transfer/test_mediation_fee.py]: Oh I was not seeing it's capacity==1 token wei.

File: raiden/transfer/mediated_transfer/mediation_fee.py (hunk 3 above)
Comment (pirapira) [reply raiden/transfer/mediated_transfer/mediation_fee.py]: Yes, I just felt a bit random.  But I guess the callers are dealing with `None` reasonably.

File: raiden/transfer/mediated_transfer/mediation_fee.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (palango) [reply raiden/transfer/mediated_transfer/mediation_fee.py]: I changed the implementation to use `round`.

**Pull request merged**